plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -60,120 +64,110 @@ becomes
-            let body = varBody varInfo
-                defAsInlineTerm = varDef varInfo
-                inlineTermToTerm :: InlineTerm tyname name uni fun ann
-                  -> Term tyname name uni fun ann
-                inlineTermToTerm (Done (Dupable var)) = var
-                def = inlineTermToTerm defAsInlineTerm
-                fullyApplied = isFullyApplied (arity varInfo) ctx
-                -- It is the body that we will be left with in the program after we have
-                -- reduced the saturated application, so the size increase we will be left
-                -- with comes from the body, and that is what we need to check is okay
-                bodySizeOk = sizeIsAcceptable body
+-- | Fully apply the RHS of the given variable to the given arguments, and beta-reduce
+-- the application, if possible.
+fullyApplyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The variable
+  VarInfo tyname name uni fun ann ->
+  -- | The arguments
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+fullyApplyAndBetaReduce info = go (varBody info) (varArity info)
+  where
+    go ::
+      Term tyname name uni fun ann ->
+      Arity tyname name ->
+      AppContext tyname name uni fun ann ->
+      InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+    go acc arity args = case (arity, args) of
+      -- success
+      ([], _) -> pure . Just $ fillAppContext acc args
+      (TermParam param : arity', TermAppContext arg _ args') ->
+        ifM
+          (safeToSubst param arg)
+          ( go
+              (termSubstNames (\n -> if n == param then Just arg else Nothing) acc)
+              arity'
+              args'
+          )
+          (pure Nothing)
+      (TypeParam param : arity', TypeAppContext arg _ args') ->
+        go
+          (termSubstTyNames (\n -> if n == param then Just arg else Nothing) acc)
+          arity'
+          args'
+      _ -> pure Nothing
+
+    -- Checks whether it is safe (i.e., preserves semantics) to perform beta reduction,
+    -- namely, turning `(\a -> body) arg` into `body [a := arg]`.
+    safeToSubst :: name -> Term tyname name uni fun ann -> InlineM tyname name uni fun ann Bool
+    safeToSubst n = effectSafe (varBody info) Strict n <=< checkPurity
+
+-- | Consider whether to inline an application.
+inlineSaturatedApp ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  Term tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Term tyname name uni fun ann)
+inlineSaturatedApp t
+  | (Var _ann name, args) <- splitApplication t =
+      gets (lookupVarInfo name) >>= \case
+        Just varInfo -> fullyApplyAndBetaReduce varInfo args >>= \case
+          Just fullyApplied -> do
+            let def = varDef varInfo
+                -- Inline only if the size is no bigger than not inlining.
```

Comment:
Added the following:

```
The number of AST nodes is an approximate rather than a precise measure. It correlates,
but doesn't directly map to the size of the serialised script. Different kinds of AST nodes
may take different number of bits to encode; in particular, a `Constant` node always
counts as one AST node, but the constant in it can be of arbitrary size. Then, would it be
better to use the size of the serialised term, instead of the number of AST nodes? Not
necessarily, because other transformations, such as CSE, may change the size further;
specifically, if a large constant occurs multiple times in a program, it may get CSE'd.

In general there's no reliable way to precisely predict the size impact of an inlining
decision, and we believe the number of AST nodes is in fact a good approximation.
```